src/dependency-manager/src/configs/service-config.ts

```diff
@@ -0,0 +1,292 @@
+
+  abstract getDebugVolumes(): Map<string, BaseVolumeConfig>;
+  abstract setDebugVolumes(volumes: Map<string, BaseVolumeConfig>): void;
+
+  /**
+   * Retrieve the fully resolvable service ref (e.g. <account>/<name>:<tag>) of the service
+   */
+  public getResolvableRef(): string {
+    const ref = this.getRef();
+    const name = this.getName();
+
+    if (ref && /^[a-zA-Z0-9-_]+\/[a-zA-Z0-9-_]+\:[a-zA-Z0-9-_]+$/.test(ref)) {
+      return ref;
+    } else if (ref && name) {
+      return `${name}:${ref}`;
+    } else if (name) {
+      return name;
+    }
+
+    throw new Error('A name or ref is required on all services');
+  }
+
+  /**
+   * Get all the volumes for the service
+   * @param debug [boolean] whether or not to include debug volumes (default: false)
+   */
+  public getResolvableVolumes(debug = false): Map<string, BaseVolumeConfig> {
+    const res = this.getVolumes();
+    if (debug) {
+      return new Map([
+        ...res,
+        ...this.getDebugVolumes(),
+      ]);
+    }
+
+    const debug_path = this.getDebugPath();
+    if (debug_path) {
+      res.forEach((value, key) => {
+        if (value.host_path) {
+          value.host_path = path.join(debug_path, value.host_path);
+          res.set(key, value);
+        }
+      });
+    }
+
+    return res;
+  }
+
+  /**
+   * Combine the build context with the debug path (if it exists) to consolidate
+   * the object for build-time operations.
+   */
+  public getResolvableBuildConfig(): BaseBuildConfig | undefined {
+    let debug_path = this.getDebugPath();
+    if (debug_path) {
+      const debug_path_lstat = fs.lstatSync(debug_path);
+      if (debug_path_lstat.isFile()) {
+        debug_path = path.dirname(debug_path);
+      }
+    }
+
+    let build_config = this.getBuildConfig();
+
+    if (!debug_path && !build_config || this.getImage()) {
+      return undefined;
+    }
+
+    build_config = build_config || {};
+    if (build_config.context && debug_path) {
+      build_config.context = path.join(debug_path, build_config.context);
+    } else if (debug_path) {
+      build_config.context = debug_path;
+    }
+
+    if (build_config.dockerfile && debug_path) {
+      build_config.dockerfile = path.join(debug_path, build_config.dockerfile);
+    }
+
+    return build_config;
+  }
+
+  /**
+   * Get either the debug or default command to run
+   * @param debug [boolean] (default: false)
+   */
+  public getPrioritizedCommand(debug = false) {
+    if (debug) {
+      return this.getDebugCommand() || this.getCommand();
+    }
+
+    return this.getCommand();
+  }
+
+  /**
+   * Get either the debug or default entrypoint to run
+   * @param debug [boolean] (default: false)
+   */
+  public getPrioritizedEntrypoint(debug = false) {
+    if (debug) {
+      return this.getDebugEntrypoint() || this.getEntrypoint();
+    }
+
+    return this.getEntrypoint();
+  }
+
+  public merge(config: BaseServiceConfig) {
+    this.setRef(config.getRef() || this.getRef());
```

Comment:
I don't understand the explicit merge if we use class-transformer in the constructor to merge in config `plainToClassFromExist`. Much preferred having class-transformer merge because it lowers the surface area for unwanted changes and we never have to worry about forgetting to include new flags.

Reply:
More of a question of why did we stop using class-transformer if we aren't going to remove it entirely.